### PR TITLE
Consider road slope in ego vehicle simulation

### PR DESCRIPTION
**Pull request by HansRobo**
# Description

## Abstract

We started to consider road slope in lanelet2 map.
In this pull-request, we consider about gravity effect at road slope in ego vehicle simulation.

## Background

In horizontal road, gravity has no effect to vehicle behavior.
But in the sloped road, the gravity has some horizontal component to the road like the figure below. 
image

This gravity component effects to the vehicle simulation behavior especially like acceleration and deceleration.

### The algorithm to consider sloped road

1. Do lane-matching and get s value in the lane coordinate
2. Calculate road slope angle by centerpoint of the lane at matched s value.
3. Calculate the road horizontal component of gravity
4. Apply the acceleration calculated in step 3 to the longitudinal acceleration input of the vehicle model.

### `consider_acceleration_by_road_slope` flag to toggle function

It is togglable to enable and disable the consideration of sloped road in ego vehicle simulation by using `consider_acceleration_by_road_slope` flag.
The flag can be set in arguments of `ros2 launch` like below.

```bash
ros2 launch scenario_test_runner scenario_test_runner.launch.py \
  architecture_type:=awf/universe \
  record:=false \
  scenario:='$(find-pkg-share scenario_test_runner)/scenario/sample.yaml' \
  sensor_model:=sample_sensor_kit \
  vehicle_model:=sample_vehicle \
  consider_acceleration_by_road_slope:=true
```

To ensure backward-compatibility, this flag is set to `false` defaultly.

### Subsequent changes

This pull-request does not implement all consideration to sloped road.
I am implementing the remaining changes on https://github.com/tier4/scenario_simulator_v2/pull/1103.
By the way, the flag name `consider_acceleration_by_road_slope` is counterpart to `consider_pose_by_road_slope`, which will be introduced in #1103.

### pull-requests that were the basis for the change

- https://github.com/autowarefoundation/autoware.universe/pull/4941
- https://github.com/autowarefoundation/autoware.universe/pull/5199
- https://github.com/autowarefoundation/autoware.universe/pull/5501
- https://github.com/autowarefoundation/autoware.universe/pull/5616

# Destructive Changes

There is no destructive changes because default behavior has not changed.

The regression test are done with default flag ( [INTERNAL LINK](https://github.com/tier4/sim_evaluation_tools/issues/245) )

# Known Limitations

In this pull-request, we did not considered about lane lateral slope and lateral accelerration effect to vehicle model. 
If `simple_planning_simulator` in `autoware.universe` started to support them, `scenario_simulator_v2` will start that too.

### CODE REVIEW

**Comment by HansRobo** [comment]
Waiting for #1184

File: simulation/simple_sensor_simulator/src/vehicle_simulation/ego_entity_simulation.cpp [hunk 1]
```diff
@@ -108,8 +109,8 @@ auto EgoEntitySimulation::makeSimulationModel(
   const auto steer_time_delay           = getParameter<double>("steer_time_delay",            0.24);
   const auto vel_lim                    = getParameter<double>("vel_lim",                     parameters.performance.max_speed);  // 50.0
   const auto vel_rate_lim               = getParameter<double>("vel_rate_lim",                parameters.performance.max_acceleration);  // 7.0
-  const auto vel_time_constant          = getParameter<double>("vel_time_constant",           0.1);
-  const auto vel_time_delay             = getParameter<double>("vel_time_delay",              0.1);
+  const auto vel_time_constant          = getParameter<double>("vel_time_constant",           0.1);  // 0.5 is default value on simple_planning_simulator
+  const auto vel_time_delay             = getParameter<double>("vel_time_delay",              0.1);  // 0.25 is default value on simple_planning_simulator
```
Comment (hakuturu583) [inline simulation/simple_sensor_simulator/src/vehicle_simulation/ego_entity_simulation.cpp:113]:
```suggestion
  const auto vel_time_constant          = getParameter<double>("vel_time_constant",           0.1);  // @note 0.5 is default value on simple_planning_simulator
  const auto vel_time_delay             = getParameter<double>("vel_time_delay",              0.1);  // @note 0.25 is default value on simple_planning_simulator
```

File: simulation/simple_sensor_simulator/src/vehicle_simulation/ego_entity_simulation.cpp [hunk 2]
```diff
@@ -297,6 +311,90 @@ void EgoEntitySimulation::update(
   updatePreviousValues();
 }
 
+auto EgoEntitySimulation::getMatchedLaneletPoseFromEntityStatus(
+  const traffic_simulator_msgs::msg::EntityStatus & status, const double entity_width) const
+  -> std::optional<traffic_simulator_msgs::msg::LaneletPose>
+{
+  // the lanelet matching algorithm should be equivalent to the one used in EgoEntity::setMapPose
```
Comment (hakuturu583) [inline simulation/simple_sensor_simulator/src/vehicle_simulation/ego_entity_simulation.cpp:318]:
```suggestion
  // @note The lanelet matching algorithm should be equivalent to the one used in EgoEntity::setMapPose
```

File: simulation/simple_sensor_simulator/src/vehicle_simulation/ego_entity_simulation.cpp [hunk 3]
```diff
@@ -297,6 +311,90 @@ void EgoEntitySimulation::update(
...
+    status_, std::max(
+               vehicle_parameters.axles.front_axle.track_width,
+               vehicle_parameters.axles.rear_axle.track_width));
+  if (not ego_lanelet) {
+    return 0.0;
+  }
+
+  auto centerline_points = hdmap_utils_ptr_->getCenterPoints(ego_lanelet.value().lanelet_id);
+
+  // copied from motion_util::findNearestSegmentIndex
```
Comment (hakuturu583) [inline simulation/simple_sensor_simulator/src/vehicle_simulation/ego_entity_simulation.cpp:352]:
```suggestion
  // @note Copied from motion_util::findNearestSegmentIndex
```

Comment (hakuturu583) [reply simulation/simple_sensor_simulator/src/vehicle_simulation/ego_entity_simulation.cpp:352]: Please show the permalink of the Autoware code you referenced using the @sa command in Doxygen.

File: simulation/simple_sensor_simulator/src/vehicle_simulation/ego_entity_simulation.cpp [hunk 4]
```diff
@@ -297,6 +311,90 @@ void EgoEntitySimulation::update(
...
+
+  geometry_msgs::msg::Point ego_point;
+  ego_point.x = vehicle_model_ptr_->getX();
+  ego_point.y = vehicle_model_ptr_->getY();
+  const size_t ego_seg_idx = find_nearest_segment_index(centerline_points, ego_point);
+
+  const auto & prev_point = centerline_points.at(ego_seg_idx);
+  const auto & next_point = centerline_points.at(ego_seg_idx + 1);
+
+  // calculate ego yaw angle on lanelet coordinates
```
Comment (hakuturu583) [inline simulation/simple_sensor_simulator/src/vehicle_simulation/ego_entity_simulation.cpp:384]:
```suggestion
  // @note Calculate ego yaw angle on lanelet coordinates
```

File: simulation/simple_sensor_simulator/src/vehicle_simulation/ego_entity_simulation.cpp [hunk 5]
```diff
@@ -297,6 +311,90 @@ void EgoEntitySimulation::update(
...
+  const double lanelet_yaw = std::atan2(next_point.y - prev_point.y, next_point.x - prev_point.x);
+  const double ego_yaw_against_lanelet = vehicle_model_ptr_->getYaw() - lanelet_yaw;
+
+  // calculate ego pitch angle considering ego yaw.
```
Comment (hakuturu583) [inline simulation/simple_sensor_simulator/src/vehicle_simulation/ego_entity_simulation.cpp:388]:
```suggestion
  // @note calculate ego pitch angle considering ego yaw.
```

File: simulation/traffic_simulator/src/entity/ego_entity.cpp [hunk 6]
```diff
@@ -275,6 +275,8 @@ auto EgoEntity::setMapPose(const geometry_msgs::msg::Pose & map_pose) -> void
 {
   const auto unique_route_lanelets = traffic_simulator::helper::getUniqueValues(getRouteLanelets());
   std::optional<traffic_simulator_msgs::msg::LaneletPose> lanelet_pose;
+
+  // the lanelet matching algorithm should be equivalent to the one used in EgoEntitySimulation::getMatchedLaneletPoseFromEntityStatus
```
Comment (hakuturu583) [inline simulation/traffic_simulator/src/entity/ego_entity.cpp:279]:
```suggestion
  // @note The lanelet matching algorithm should be equivalent to the one used in EgoEntitySimulation::getMatchedLaneletPoseFromEntityStatus
```

**Review by hakuturu583 — CHANGES_REQUESTED** [review]

**Review by hakuturu583 — APPROVED** [review]

**Pull request merged**